app/src/main/java/org/mozilla/tv/firefox/toolbar/ToolbarViewModel.kt

```diff
@@ -0,0 +1,183 @@
+        val pinChecked: Boolean,
+        val turboChecked: Boolean,
+        val desktopModeEnabled: Boolean,
+        val desktopModeChecked: Boolean,
+        val urlBarText: String
+    )
+
+    private var previousURLHost: String? = null
+
+    // Values should be pushed to _events using setValue. Two values are set in
+    // rapid succession using postValue, only the latest will be received
+    private var _events = MutableLiveData<Consumable<BrowserNavigationOverlay.Action>>()
+    val events: LiveData<Consumable<BrowserNavigationOverlay.Action>> = _events
+
+    val state: LiveData<ToolbarViewModel.State> =
+        LiveDataHelper.combineLatest(sessionRepo.state, pinnedTileRepo.getPinnedTiles()) { sessionState, pinnedTiles ->
+
+            // The menu back button should not be enabled if the previous screen was our initial url (home)
+            fun isBackEnabled() = sessionState.backEnabled && sessionState.currentBackForwardIndex > 1
+            fun currentUrlIsPinned() = pinnedTiles.containsKey(sessionState.currentUrl)
+
+            ToolbarViewModel.State(
+                backEnabled = isBackEnabled(),
+                forwardEnabled = sessionState.forwardEnabled,
+                refreshEnabled = !sessionState.currentUrl.isEqualToHomepage(),
+                pinEnabled = !sessionState.currentUrl.isEqualToHomepage(),
+                pinChecked = currentUrlIsPinned(),
+                turboChecked = turboMode.isEnabled(),
+                desktopModeEnabled = !sessionState.currentUrl.isEqualToHomepage(),
+                desktopModeChecked = sessionState.desktopModeActive,
+                urlBarText = UrlUtils.toUrlBarDisplay(sessionState.currentUrl)
+            )
+        }
+
+    init {
+        disableDesktopModeWhenHostChanges()
+        showOverlayWhenUrlIsHome()
+    }
+
+    @UiThread
+    fun backButtonClicked() {
+        sessionRepo.exitFullScreenIfPossibleAndBack()
+        setOverlayVisible(false)
+    }
+
+    @UiThread
+    fun forwardButtonClicked() {
+        sessionRepo.goForward()
+        setOverlayVisible(false)
+    }
+
+    @UiThread
+    fun reloadButtonClicked() {
+        sessionRepo.reload()
+        sessionRepo.pushCurrentValue()
+        setOverlayVisible(false)
+    }
+
+    @UiThread
+    fun pinButtonClicked() {
+        val pinChecked = state.value?.pinChecked ?: return
+        val url = sessionRepo.state.value?.currentUrl ?: return
+
+        sendOverlayClickTelemetry(NavigationEvent.PIN_ACTION, pinChecked = !pinChecked)
+
+        if (pinChecked) {
+            pinnedTileRepo.removePinnedTile(url)
+            _events.value = Consumable.from(BrowserNavigationOverlay.Action.ShowTopToast(R.string.notification_unpinned_site))
+        } else {
+            pinnedTileRepo.addPinnedTile(url, sessionRepo.currentURLScreenshot())
+            _events.value = Consumable.from(BrowserNavigationOverlay.Action.ShowTopToast(R.string.notification_pinned_site))
+        }
+        setOverlayVisible(false)
```

Comment:
Yeah, that is correct @psymoon . But I see where your doubt is coming from. Unfortunately we can only make an assumption here but it would be interesting to see how many users stay on the website after pinning and how many re-open the overlay. And base on that either change or leave the behavior as it is.